src/inference.py

```diff
@@ -0,0 +1,128 @@
+parser.add_argument("--img_height", type=int, default=128, help="size of image height")
+parser.add_argument("--img_width", type=int, default=128, help="size of image width")
+parser.add_argument("--channels", type=int, default=1, help="number of image channels")
+parser.add_argument("--n_downsample", type=int, default=2, help="number downsampling layers in encoder")
+parser.add_argument("--dim", type=int, default=32, help="number of filters in first encoder layer")
+
+opt = parser.parse_args()
+print(opt)
+
+os.makedirs('out_infer', exist_ok=True)
+
+cuda = True if torch.cuda.is_available() else False
+Tensor = torch.cuda.FloatTensor if cuda else torch.Tensor
+input_shape = (opt.channels, opt.img_height, opt.img_width)
+
+# Dimensionality (channel-wise) of image embedding
+shared_dim = opt.dim * 2 ** opt.n_downsample
+
+# Initialize encoder and decoder
+encoder = Encoder(dim=opt.dim, in_channels=opt.channels, n_downsample=opt.n_downsample)
+G = Generator(dim=opt.dim, out_channels=opt.channels, n_upsample=opt.n_downsample, shared_block=ResidualBlock(features=shared_dim))
+
+if cuda:
+    encoder = encoder.cuda()
+    G = G.cuda()
+
+assert os.path.exists("saved_models/%s/encoder_%02d.pth" % (opt.model_name, opt.epoch)), 'Check that trained encoder exists'
+assert os.path.exists("saved_models/%s/G%s_%02d.pth" % (opt.model_name, opt.gen_id, opt.epoch)), 'Check that trained generator exists'
+    
+# Load pretrained models
+encoder.load_state_dict(torch.load("saved_models/%s/encoder_%02d.pth" % (opt.model_name, opt.epoch)))
+G.load_state_dict(torch.load("saved_models/%s/G%s_%02d.pth" % (opt.model_name, opt.gen_id, opt.epoch)))
+
+# Set to eval mode 
+encoder.eval()
+G.eval()
+
+# Load audio and preprocess
+sample = preprocess_wav(opt.wav)
+spect_src = melspectrogram(sample)
+
+# --------------------
+#  Local Inference
+# --------------------
+
+def infer(S):
+    """Takes in a standard sized spectrogram, returns converted version"""
+    S = torch.from_numpy(S)
+    S = S.view(1, 1, opt.img_height, opt.img_width)
+    
+    X = Variable(S.type(Tensor))
+    mu, Z = encoder(X)  # Get shared latent representation
+    fake_X = G(Z)  # Translate speech
+    return to_numpy(fake_X)
+
+# ------------------------------------------
+#  Global Inference (w/ a sliding window)
+# ------------------------------------------
+
+spect_src = np.pad(spect_src, ((0,0),(opt.img_width,opt.img_width)), 'constant')  # padding for consistent overlap
+length = spect_src.shape[1]
+spect_trg = np.zeros(spect_src.shape)
+hop = opt.img_width // opt.n_overlap
+
+for i in tqdm(range(0, length, hop)):
+    x = i + opt.img_width
+         
+    # Get cropped spectro of right dims
+    if x <= length:
+        S = spect_src[:, i:x]
+    else:  # pad sub on right if includes sub segments out of range
+        S = spect_src[:, i:]
+        S = np.pad(S, ((0,0),(x-length,0)), 'constant') 
+        
+    T = infer(S) # perform inference from trained model
```

Comment:
This is a little bit time consuming to do the inference on a single window at a time, it is better to do it on all windows at once and then do the post-processing after that, this should be much faster.

Reply:
Honestly I don't think it's entirely necessary as the inference part is the fastest part of the code (in training and eval its the data_proc that slows things down)

As it is right now, it globally infers at 2.24 iterations/second.

Other than that, if we make it depend on all windows at once - this will make things more limiting for lower resource users (say CPU only users). And also if an audio file is quite long (say an hour long), a lot of people could easily run out of memory.

What's your take on my reasoning?

Reply:
I still think it should be faster for batch computation of spectrograms but it could be not that much of improvement since this code is processing one sample at a time. lets keep it as it is for now then.

Reply:
If the inference was for a large set of data rather than one file, I think it would make much more of a difference to be fair. So at least batch computation is something to keep in mind for that (if we ever do that)